File: evals/refs/pq.py
import sys

class Heap:
    def __init__(self):
        self.h = []
    def push(self, v):
        self.h.append(v)
        i = len(self.h) - 1
        while i > 0:
            p = (i - 1) // 2
            if self.h[p] > self.h[i]:
                self.h[p], self.h[i] = self.h[i], self.h[p]
                i = p
            else:
                break
    def pop(self):
        if not self.h: return None
        top = self.h[0]
        last = self.h.pop()
        if self.h:
            self.h[0] = last
            i = 0
            n = len(self.h)
            while True:
                l = 2*i + 1; r = 2*i + 2
                m = i
                if l < n and self.h[l] < self.h[m]: m = l
                if r < n and self.h[r] < self.h[m]: m = r
                if m != i:
                    self.h[m], self.h[i] = self.h[i], self.h[m]
                    i = m
                else:
                    break
        return top

lines = sys.stdin.read().splitlines()
idx = 0
T = int(lines[idx]); idx += 1
out = []
for _ in range(T):
    Q = int(lines[idx]); idx += 1
    h = Heap()
    for _ in range(Q):
        op = lines[idx]; idx += 1
        if op.startswith('p '):
            v = int(op[2:])
            h.push(v)
        else:
            r = h.pop()
            out.append(str(r) if r is not None else '-')
sys.stdout.write('\n'.join(out) + '\n')

File: evals/refs/pq.rs
use std::io::{self, BufRead, Write, BufWriter};

struct Heap { h: Vec<i64> }
impl Heap {
    fn push(&mut self, v: i64) {
        self.h.push(v);
        let mut i = self.h.len() - 1;
        while i > 0 {
            let p = (i - 1) / 2;
            if self.h[p] > self.h[i] { self.h.swap(p, i); i = p; } else { break; }
        }
    }
    fn pop(&mut self) -> Option<i64> {
        if self.h.is_empty() { return None; }
        let top = self.h[0];
        let last = self.h.pop().unwrap();
        if !self.h.is_empty() {
            self.h[0] = last;
            let mut i = 0usize;
            let n = self.h.len();
            loop {
                let l = 2*i + 1; let r = 2*i + 2;
                let mut m = i;
                if l < n && self.h[l] < self.h[m] { m = l; }
                if r < n && self.h[r] < self.h[m] { m = r; }
                if m != i { self.h.swap(m, i); i = m; } else { break; }
            }
        }
        Some(top)
    }
}

fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut lines = stdin.lock().lines();
    let t: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
    for _ in 0..t {
        let q: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
        let mut h = Heap { h: Vec::new() };
        for _ in 0..q {
            let line = lines.next().unwrap().unwrap();
            if line.starts_with("p ") {
                let v: i64 = line[2..].trim().parse().unwrap();
                h.push(v);
            } else {
                match h.pop() {
                    Some(v) => writeln!(out, "{}", v).unwrap(),
                    None => writeln!(out, "-").unwrap(),
                }
            }
        }
    }
}
